SIW Gen3 - Document afterRender revert (#4784)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/custom-widget-gen3/main/index.md

   > **Note:** To discard your changes without publishing them, click **Revert changes** or click the **Code editor** toggle again. Turning off the code editor restores the default JavaScript code.

### Use afterRender with the third generation

> The third generation of the Sign-In Widget is built on [Preact](https://preactjs.com/), a lightweight React alternative. This means that the `afterRender` function doesn't work with the third generation.

Reviewer: The `afterRender` function is still triggered, and useful for logging and "pure" callbacks. They just won't work with DOM manipulation and similar render-related changes.

```suggestion
The third generation of the Sign-In Widget is built on [Preact](https://preactjs.com/), a lightweight React alternative.
```

> The third generation of the Sign-In Widget is built on [Preact](https://preactjs.com/), a lightweight React alternative. This means that the `afterRender` function doesn't work with the third generation.

Reviewer 2: ```suggestion
The third generation of the Sign-In Widget is built on [Preact](https://preactjs.com/), a lightweight React alternative. This means that the `afterRender` function will not work when used to make DOM manipulations in the third generation.
```

With the second generation of the Sign-In Widget, you can use [`afterRender`] (https://github.com/okta/okta-signin-widget?tab=readme-ov-file#afterrender)to change label text or input labels on a page.

With the third generation, `afterRender` is incompatible with customizations that make DOM manipulations and other render-related side-effects (see [Components and Hooks must be pure](https://react.dev/reference/rules/components-and-hooks-must-be-pure)).

#### Resolve the afterRender revert

> To keep the third generation of the Sign-In Widget from reverting your `afterRender` customizations, use the DOM `MutationObserver()` function. See [MutationObserver](https://dom.spec.whatwg.org/#ref-for-dom-mutationobserver-mutationobserver).

Reviewer: Changed the reference link URL. Feel free to keep the original if providing the ref is standard. The MDN link is more useful.

```suggestion
To keep the third generation of the Sign-In Widget from reverting your `afterRender` customizations, use the `MutationObserver()` interface. See [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).
```

Author: Providing the ref is standard.

> **Note:** Use the `MutationObserver` with caution. The following solutions are only for developers with experience using this function.

To update UI elements, consider the following example:

```javascript
<script type="text/javascript" nonce="{{nonceValue}}">
    var config = OktaUtil.getSignInWidgetConfig();

    var oktaSiwRoot = document.querySelector('#okta-login-container');
    // The following allows you to reference the context from each render

>     let contextObj = {};

Reviewer: ```suggestion
    var contextObj = {};
```

    function cb(mutations, observer) {
      // For the primary auth form, updates the button label
      if (contextObj.formName === 'identify') {

>         const el = document.querySelector('[data-type="save"]');

Reviewer: ```suggestion
        var el = document.querySelector('[data-type="save"]');
```

        if (el) { el.textContent = 'Some new label'; }
      }
      // For the reset-authenticator view, updates the button label
      if (contextObj.formName === 'reset-authenticator') {

>         const el = document.querySelector('[data-type="save"]');

Reviewer: ```suggestion
        var el = document.querySelector('[data-type="save"]');
```

      contextObj = context;
      // The following condition only executes the observer for specific views/forms
      if (context.formName === 'identify' || context.formName === 'reset-authenticator') {
        // Pauses
        observer.disconnect();

        // Calls once after initial render
        cb();

        // Observes for re-renders
        observer.observe(oktaSiwRoot, {
          subtree: true,
          childList: true,
          attributes: true,
          characterData: true,
        });
      }
    });

 </script>
 ```

 The third generation can use the `afterRender` function for non-DOM manipulations without additional logic. The following example doesn't update the UI, but instead sends a log to your external logging service:

 ```javascript
 <script type="text/javascript" nonce="{{nonceValue}}">
    var config = OktaUtil.getSignInWidgetConfig();

    // Renders the Okta Sign-In Widget
    var oktaSignIn = new OktaSignIn(config);

    // The following varies based on your own configuration
    oktaSignIn.renderEl({ el: '#okta-login-container' }, OktaUtil.completeLogin, function (error) {
       console.log(error.message, error);
    });

>     oktaSignIn.on('afterRender', (context) => {

Reviewer: ```suggestion
    oktaSignIn.on('afterRender', function (context) {
```

   > **Note:** To discard your changes without publishing them, click **Revert changes** or click the **Code editor** toggle again. Turning off the code editor restores the default JavaScript code.

### Use afterRender with the third generation

> The third generation of the Sign-In Widget is built on React.js. It interprets the `afterRender` function in a way that's incompatible with customizations. The Sign-In Widget reverts your customizations to Okta default settings.

Reviewer: ```suggestion
The third generation of the Sign-In Widget is built on [Preact](https://preactjs.com/) -- a lightweight React alternative. This means the `afterRender` function incompatible with customizations that make DOM manipulations and other render-related side-effects (see [Components and Hooks must be pure](https://react.dev/reference/rules/components-and-hooks-must-be-pure)).
```

> Typically, with the second generation of Sign-In Widget, you can use `afterRender` to change label text or input labels on a page. The Sign-In Widget triggers `afterRender` when transitioning to a new page and animations have finished. See [afterRender](https://github.com/okta/okta-signin-widget?tab=readme-ov-file#afterrender).

Reviewer: ```suggestion
With the second generation of Sign-In Widget, you can use `afterRender` to change label text or input labels on a page. The Sign-In Widget triggers `afterRender` when transitioning to a new page and animations have finished. See [afterRender](https://github.com/okta/okta-signin-widget?tab=readme-ov-file#afterrender). This will not work in the third generation of Sign-in Widget.
```

   > **Note:** To discard your changes without publishing them, click **Revert changes** or click the **Code editor** toggle again. Turning off the code editor restores the default JavaScript code.

### About the afterRender function

The third generation of the Sign-In Widget is built on [Preact](https://preactjs.com/), a lightweight React alternative. This means that the [`afterRender`] (https://github.com/okta/okta-signin-widget?tab=readme-ov-file#afterrender) function doesn't work when used to make DOM manipulations and other render-related side effects. If you use `afterRender` for DOM manipulations, the Okta Sign-In Widget reverts any customizations to default settings. See [Components and Hooks must be pure](https://react.dev/reference/rules/components-and-hooks-must-be-pure).

* To use `afterRender` for DOM manipulations, consider using the `MutationObserver` function.
* To use `afterRender` for non-DOM manipulations, you don't need the `MutationObserver` function.

#### Use MutationObserver for DOM manipulations

> To keep the third generation of the Sign-In Widget from reverting your `afterRender` customizations, use the DOM `MutationObserver` function. See [MutationObserver](https://dom.spec.whatwg.org/#ref-for-dom-mutationobserver-mutationobserver).

Reviewer 3: ```suggestion
To prevent the third generation of the Sign-In Widget from reverting your `afterRender` customizations, use the DOM `MutationObserver` function. See [MutationObserver](https://dom.spec.whatwg.org/#ref-for-dom-mutationobserver-mutationobserver).
```
